### DISCUSSION TITLE
Introducing Tailwind Play, the official Tailwind CSS playground!

### DISCUSSION THREAD

**Discussion by adamwathan** [tailwindlabs/tailwindcss#2511] [Show and tell]
Today we're excited to release the first version of Tailwind Play, an advanced online playground for Tailwind CSS that lets you use all of Tailwind's build-time features directly in the browser.

It includes support for all of Tailwind's coolest features, plus tons of stuff that's even better in Tailwind Play than it is in your editor, like:

- Customizing your Tailwind theme
- Enabling special variants, like `group-hover` or `focus-within`
- Using custom directives in your CSS like `@apply`, `@variants`, and `@responsive`
- Adding plugins like `@tailwindcss/typography`
- Intelligent code completion and linting
- Responsive design mode
- One-click sharing

Check it out at [play.tailwindcss.com](https://play.tailwindcss.com) and let us know what you think!

**Comment by philippbosch** [comment]
This is awesome! Thanks for this!! :tada:

What would make it even better would be a way to see the resolved config (default + custom merged) somewhere. This could help new users understand how to use the config file.

**Comment by rjgux** [comment]
This is awesome! Just as we were experimenting with the idea of designing in code with Tailwind, this will help us massively in the future.

**Reply by simonswiss** [reply]
Yep - designing in code with Tailwind is definitely unlocked with this new playground! 🙌

**Comment by skttl** [comment]
How do you make the preview viewport scale like that? So awesome!

**Comment by MartinHughes-BPC** [comment]
I can only think of one response to this:

https://play.tailwindcss.com/QT0aK8Av4H?layout=vertical&file=html

**Reply by simonswiss** [reply]
Haha thank you! 🎉

**Comment by swyxio** [comment]
congrats team, i am going to use the heck out of this!

y'all probably foresaw this question coming and didn't answer it but i gotta ask anyway...

**will this be open sourced**? i'd like to take parts of it and add some things of my own.

**Reply by santicros** [reply]
It's awesome. Would love that too!

**Reply by gcavanunez** [reply]
might be wrong here but seems like when hitting the save button, it creates a new static page which is so cool. less overhead of managing a database.

**Reply by swyxio** [reply]
i could be wrong here but i reckon thats a feature of nextjs https://nextjs.org/blog/next-9-4#incremental-static-regeneration-beta

**Reply by adriandmitroca** [reply]
its just a classic Laravel API (which is curious, as in beginning it was based on AWS DynamoDB but ultimately dropped by looking at the commits)

**Reply by adamwathan** [reply]
Yep just boring API with a MySQL database. We made the switch because we were worried about a $100k AWS bill with a wide-open endpoint that writes to DynamoDB. With the Laravel API we can easily rate limit it per IP (backed by Redis), and it's all just one a small EC2 server.

The entire Laravel API is only ~10 lines of changed code from what you get when running `laravel new` 🙃

**Reply by adriandmitroca** [reply]
Doesn't DynamoDB have any rate limiting whatsoever?

Thanks for the insights though, super interesting!

**Comment by santicros** [comment]
It's very very good. Thanks all the TailwindLabs team!

One suggestion I have is: Do you think would be interesting to add something to easily switch between breakpoints? Like small buttons with sm, md, lg, xl...?

Thanks!!

**Comment by gcavanunez** [comment]
So cool how it's optimized, out of curiosity any reason why the responsive mode doesn't include the share button?

**Comment by humledumle** [comment]
Amazing!

**Comment by Natetronn** [comment]
This is great! Thank you.

A few ideas (no pressure or anything, just what came to mind the first time I saw the Playground.)

1. It would be cool to incorporate the docs in an intelligent way. If nothing else, some sort of "cheat sheet" that stays in the playground. Perhaps a search feature, like on Tailwind proper, but it opens up in a modal or slides out or something; but don't send me to the docs; instead give me the option to go there myself.
2. When creating a share link, being able to link (and have it highlight) a specific line or set of lines, like you can on Github. I think this would be helpful for people, this so they can narrow down what they are referring to or for people having trouble to share the section they are working on etc.
3. Add an embed feature. This way people who write tutorials on Tailwind can embed various examples into their own site. Could be the whole document or just a specific line or set of lines etc. (see # 2 above.) Could be used for tutorials, but maybe even cooler, could one day be used in places like Stack Exchange (think JSFiddle but, for Tailwind questions/answers.)

**Comment by JeffreyHyer** [comment]
Small bug: When I hit "Share" a link is generated. When I switch to responsive mode the state clears the link and if I hit "Share" again a new link is generated.

Edit: I don't mean to sound like I'm complaining/unappreciative because nothing could be further from the truth. I'm over the moon about this and everything else you guys are doing with Tailwind! This is amazing, keep up the great work guys! 🎉 🙌

**Reply by adamwathan** [reply]
This is by design actually if I'm understanding correctly — the "Share" button captures the current layout state, selected file, and even the responsive viewport size, so changing any of those invalidates the URL.

**Reply by emicheldev** [reply]
ok

**Reply by Natetronn** [reply]
Perhaps the layout state, selected file and responsive viewport etc. could be decoupled from the the url a bit more and become more like query strings or something. This way people can share various states of the app (think of how Youtube does time stamps.)

**Reply by adamwathan** [reply]
Already the case, take a look at the URL when you click share!

https://play.tailwindcss.com/rLhGpdKMbv?layout=horizontal&size=706x737

**Reply by Natetronn** [reply]
Ah okay, I see that now. 

Of course, now you have ~7 different URL paths for the same code, should you want to share the different views, tabs etc.

Is it not extra work/storage for your database to have so many duplicates and would it not make more sense to have the URL update dynamically, this instead of having to click share again every time one wants to share a different view?

You said it's not a bug and was designed this way but, I'm struggling to see the reasons for it being done like that. Care to enlighten me?

Note: hopefully you don't take any of this as a complaint rather, a discussion and me trying to better understand etc.

**Reply by adamwathan** [reply]
So we don't actually store 7 copies, we only store the actual code + the associated random ID. Anything in the query string is not stored in our DB, it's just added by the client at the time the snapshot is saved. If you hit the "share" button again we don't try to dedupe or reuse anything (although we certainly can if it proves to be an issue), we just create a new link.

The main reason for designing things this way is that it was easy, it works fine enough, and anything more sophisticated would take more time, which means less time to work on other hard problems we needed to solve to release it. Spending time on optimizations around storage would mean less time we could spend making third party plugins work, or getting code completion working, etc.

We had a prototype where the URL updated dynamically every time you made a change to the editor layout, but the problem was the random ID doesn't change, so it's easy to think that if you see the layout switch from vertical to horizontal in the URL that your code changes must be saved too right? But they aren't, not until you hit "share" (saving a new row to the DB on every keystroke would _really_ be wasteful).

Putting any sort of live updating in the URL implied that things were being saved, which made the user experience more confusing. Now the URL always reflects the "last save point", so if you hit cmd+R, you are taken to an exact snapshot of the last time you pressed "share".

**Reply by Natetronn** [reply]
Thanks for the detail explanation, Adam, we really appreciate it!

**Comment by emicheldev** [comment]
Awesome 🥳😍

**Comment by MartinHughes-BPC** [comment]
Further to my previous comment, I now find this to be:

https://play.tailwindcss.com/995gjq9GTW

This is going to change the way I work with Tailwind.

**Comment by daniel-norris** [comment]
Nice, I like it!

**Comment by trevyn** [comment]
This is great! Feature request: Word wrap option in the editor.

**Comment by shanerobinson** [comment]
**O... M... G...** It supports VSCode keyboard shortcuts?!?! 🤪

**Reply by Calinou** [reply]
Yes, that's because it embeds VS Code's editing core which is called [Monaco](https://microsoft.github.io/monaco-editor/). You can even right-click or press <kbd>F1</kbd> to access the command palette :slightly_smiling_face:

**Comment by graciasc** [comment]
I noticed that tailwind play doesn't support js in script tags. Will this be added in the upcoming versions ?

Ex:  https://play.tailwindcss.com/Vx0zf2TSYQ

**Comment by dgwyer** [comment]
How about adding Apline JS support? :-)

**Comment by brumm** [comment]
It's so good! I'd love to see Emmet support to help with writing lots of html :)

**Comment by ctrlShiftBryan** [comment]
Thanks for the awesome work on this! Will you be able enable the VIM keybinindings anytime soon?

**Comment by borjajimnz** [comment]
Thanks for the awesome work, but I think that Alpine must be a must!! please!!! thanks!!

**Comment by YoungElPaso** [comment]
Hi, this is an awesome tool. However, there's a few things that to me, remain confusing:
 * Ctrl+s to format is great, but a lot of editors (CodePen, VS Code, heck MS Word) use ctrl+s as the default 'Save' hotkey.  This plus the fact that several of those 'format on save' makes it seem like (in my experience) that you're 'saving' _and_ formatting when you use that hotkey.  I was convinced I'd saved some work for over an hour - and then of course lost it. This isn't helped by factors out of your control, some other online IDE'ish tools also auto-save promoting this perception, but perhaps some clarity around hot-keys would help?
 * Related, as I discovered belatedly, with this tool to 'share' is to save. I get the logic here (why save what you don't want to share?), but this also doesn't really jive with a lot of expectations ingrained from other tools.  Its an elegant solution to keeping the tool simple, but its not intuitive or clear IMO. Perhaps some help text or changing the button to 'Save+Share' or something would be useful?
* I enjoy the minimalism of the UI in general, I don't think much more is needed, but a bit of help text, and meta info about hotkeys supported, versions of TW, ability to add (or not) external libraries (thinking Alpine, jQuery etc) and just a general overview of the tool might be useful to other users. Set expectations etc.

Anyway, this is still awesome and I really need to use this to evangelize to my team and I'm a big fan of Tailwind in general. Thanks for considering the above.

**Reply by simonswiss** [reply]
This is great feedback, @YoungElPaso 👍

Thanks for sharing - we'll definitely keep this in mind as we keep iterating over Tailwind Play. Agree that the awesome idea of formatting on `ctrl + s`, especially if you have your vscode set to `formatOnSave: true`, really makes the brain think that the work was saved.

Thanks again for taking the time to share these points 👍

**Comment by qaharmdz** [comment]
This playground is a great help for beginners like me. 
If there is a suggestion that would be, similar to Github Gist, to create an account and save the work (in public or private). Thus we can search and learn about how peoples use Tailwind. I'm the type who can learn a lot more from various examples rather than reading documentation overnight :D

**Comment by jaysridhar** [comment]
Is it possible to make play.tailwind available as source or something so I can install on a local machine? I need to use other resources (such as images) on the local box and it is a pain to have to copy these over to imgur just to have it available to play.tailwind. Thanks

**Reply by Natetronn** [reply]
I think this is it here: https://github.com/tailwindlabs/play.tailwindcss.com

**Reply by dinfer** [reply]
https://github.com/tailwindlabs/play.tailwindcss.com is 404.

**Comment by ghost** [comment]
Great tool.

Two features would be nice:
1. Integration with a google doc for comments on a page.
2. A persistent url that points to the latest version of a page.

**Comment by guacamoly** [comment]
A absolutely fantastic tool. Thanks a lot!
Is there an expiration on a saved/shared link? How long do they last?

**Comment by vivekascoder** [comment]
How https://play.tailwindcss.com is so fast, like seriously it's faster then editing in vscode.

**Comment by JStuve** [comment]
Absolutely wonderful, using this for quick UI design that can easily be ported off to our application.

Quick question, how long will shared links last?

**Reply by adamwathan** [reply]
Thank you! They don't expire, so they will be live for as long as we keep paying to keep the servers running 👍🏻

**Comment by tian-yan** [comment]
It is a super helpful tool. However, can you share the backend components of this tool (version of tailwindcss, postcss-cli, autoprefixer etc)? As I came across a situation where same codes worked in Tailwind Playground, but fails in real world production deployment (css file produced by `npm run build` command), [detailed symptoms are here](https://github.com/tailwindlabs/tailwindcss/discussions/5330).  Hence I want to compare the backend components in my local development environments and in Tailwind Playground to find out differences.

**Comment by jk-jatri** [comment]
How long after the shared code from https://play.tailwindcss.com will expire?

**Reply by JStuve** [reply]
Was able to get an answer to this question few comments above; indefinitely. 


> Thank you! They don't expire, so they will be live for as long as we keep paying to keep the servers running 👍🏻

**Comment by jasonrberk** [comment]
how do I add some vanilla JS so I can build out a few on-click events?

my current options are:

1. codepen with tailwind v2
2. play w/o javascript

even vanilla JS would be tolerable :-)

**Comment by vnphanquang** [comment]
Hi there everybody thanks for all the hard work here. Is there a current way or planned support for loading a tailwind plugin at a specific commit / version? This is useful to test a plugin support before releasing it. Right now I had to do it in local dev but it would be good to have the ability to test in tailwind play production.

**Reply by vnphanquang** [reply]
A fellow developer just reminded me that tailwind play DOES in fact support this:

```javascript
module.exports = {
  plugins: [
    require('daisyui@2.3.6'),
  ],
}
```

Leaving this here in case someone needs the answer.

**Comment by mike1701** [comment]
Very nice tool for playing around or building some cool components. Thanks for so far.
But a nice feature would be, when the Dark-Mode can be apply to the rendered view for testing dark:-Modifiers.

**Comment by ErwinAI** [comment]
@adamwathan Would it be possible to look into pre-fill support? An API that allows making a new instance, prefilling it with html/css and obtaining a URL for that instance would be very useful to me. Currently, the only option available for this is codepen, but I'd much rather use Tailwind Play.

**Comment by adriatic** [comment]
This is a repost of my question in **[Discord](https://discord.com/channels/674983035946663957/674983036374220800/1094790134936915988)**, conducted here for readability reason.

I took the complete code (.jsx) from the very first [navbar in the collection of navbars](https://tailwindui.com/components/application-ui/navigation/navbars) and pasted it in https://play.tailwindcss.com/. The outcome is very different from the [originalpreview]( https://tailwindui.com/components/application-ui/navigation/navbars). By very different I mean that instead of getting this

<img width="815" alt="image" src="https://user-images.githubusercontent.com/2712405/230913840-98fc5cc6-dada-421c-a67b-6b98a62319ec.png">

I am getting this (this screenshot is cut short at the bottom)

<img width="1492" alt="image" src="https://user-images.githubusercontent.com/2712405/230914042-4e9dbbc6-b433-48ae-9c60-a63713659e3d.png">

Do I need to configure https://play.tailwindcss.com/ before pasting this code?

**Reply by vnphanquang** [reply]
I could be wrong, but last time I used tailwindplay it only supported vanilla HTML syntax and not react.

And I imagine the code snippet in your "[navbar in the collection of navbars](https://tailwindui.com/components/application-ui/navigation/navbars)" reference is only for demonstrating purposes.

**Reply by adamwathan** [reply]
Hey! As @vnphanquang mentioned, Tailwind Play doesn't support React, it's just for playing with Tailwind with vanilla HTML.

The Tailwind UI snippet you are playing with will work though if you paste it into a proper React project 👍

**Reply by adriatic** [reply]
@adamwathan If there is just one person with authority in this context, it is you 😄 - thanks for clarifying this and apologies to @vnphanquang for doubting him. 

I misread the galleries exhibited in https://tailwindui.com (which can generate .jsx files) with the belief that Tailwind Play know how to process them (not that I would be disappointed if it could 😄)

**Comment by snimavat** [comment]
It would be great if it allowed me to see list of all the play snippets created by me.
that way i can play with tailwind and come back to an older snippet whenever i need to play with it again
i dont have to recode or remember the url

**Comment by atengberg** [comment]
Is there really no "explorer" mode to see all the inspired creations?

**Comment by makikata** [comment]
Is there an way to host play ourselves? There was a public repo some time ago

<blockquote class="twitter-tweet"><p lang="en" dir="ltr">Just made the repo public if you want to poke around! 🧐<a href="https://t.co/KDMALxZMMO">https://t.co/KDMALxZMMO</a></p>&mdash; Adam Wathan (@adamwathan) <a href="https://twitter.com/adamwathan/status/1313873343590916097?ref_src=twsrc%5Etfw">October 7, 2020</a></blockquote>

**Comment by intrnl** [comment]
can we get an option for disabling text wrapping? I don't like the recent update to enable text wrapping.

**Comment by HappyLDE** [comment]
Hi please open source it again, I want to offer my clients the possibility to create tailwind components in realtime and could use the help of the source code

**Comment by scottromack** [comment]
Can we add branching?

**Comment by wladi3** [comment]
Why did you removed the source from Github? Is there any possibility to open the code? THX !!!

**Comment by inwardmovement** [comment]
When we select an intellisense suggestion, the app shows a preview of it.

How is this preview thing achieved? How can we get the same behavior in VSCode?

**Comment by Dan1ve** [comment]
A bit of a bummer that the source repo has been removed. I used this locally to fiddle around with designs and a spotty internet connection. 

To everyone looking for the source, I found this: https://codesandbox.io/p/devbox/play-tailwindcss-com-jvnel (don't know how recent this is, but it's as least something to start with)

**Comment by vortexprime24** [comment]
It is great, all I wish is JavaScript support. And, a AI integration: maybe your own? Last, collab features and desktop app, and it is ultimate

**Comment by justrach** [comment]
Hi is there any way to get the source code for the initial playground up!

**Comment by tresorama** [comment]
https://github.com/tailwindlabs/tailwindcss/discussions/13058#discussioncomment-10274721

**Comment by krabhi1** [comment]
How does it work?
